File: ci/local.py
"""Local CI provider for testing without external CI systems."""

from __future__ import annotations

import asyncio
import logging
import subprocess
import uuid
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Optional, Callable, Awaitable, Any

from .interface import (
    CIProvider,
    BuildStatus,
    BuildStatusType,
    PRInfo,
    PRStatusType,
    CIEvent,
)
from .events import CIEventEmitter, CIEventType

logger = logging.getLogger(__name__)


@dataclass
class LocalBuild:
    """Represents a local CI build."""
    run_id: str
    branch: str
    command: str
    status: BuildStatusType = BuildStatusType.PENDING
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    output: str = ""
    error: str = ""
    return_code: Optional[int] = None


@dataclass
class LocalPR:
    """Represents a local PR (simulated)."""
    pr_number: int
    source_branch: str
    target_branch: str
    title: str
    body: str = ""
    status: PRStatusType = PRStatusType.OPEN
    created_at: datetime = field(default_factory=datetime.now)
    merged_at: Optional[datetime] = None
    checks_passed: bool = False


class LocalCIProvider(CIProvider):
    """
    Local CI provider that runs tests locally.

    This is useful for:
    - Testing the orchestration system without external CI
    - Development and debugging
    - Environments without CI access
    """

    def __init__(
        self,
        repo_dir: Path,
        test_command: str = "pytest",
        build_command: str = "python -m py_compile",
        event_emitter: Optional[CIEventEmitter] = None,
    ):
        self.repo_dir = Path(repo_dir)
        self.test_command = test_command
        self.build_command = build_command
        self.event_emitter = event_emitter or CIEventEmitter()

        self._builds: Dict[str, LocalBuild] = {}
        self._prs: Dict[int, LocalPR] = {}
        self._next_pr_number = 1
        self._subscribers: List[Callable[[CIEvent], Awaitable[None]]] = []

    async def trigger_build(
        self,
        branch: str,
        config: Optional[Dict[str, Any]] = None
    ) -> str:
        """Trigger a local build/test run."""
        run_id = str(uuid.uuid4())[:8]

        # Determine command based on config
        command = config.get("command") if config else None
        if not command:
            command = self.test_command

        build = LocalBuild(
            run_id=run_id,
            branch=branch,
            command=command,
        )
        self._builds[run_id] = build

        # Run the build asynchronously
        asyncio.create_task(self._run_build(build))

        logger.info(f"Triggered local build {run_id} for {branch}")
        return run_id

    async def _run_build(self, build: LocalBuild):
        """Execute a build in the background."""
        build.status = BuildStatusType.RUNNING
        build.started_at = datetime.now()

        # Emit build started event
        await self.event_emitter.emit(CIEvent(
            event_id=str(uuid.uuid4()),
            event_type=CIEventType.BUILD_STARTED.value,
            branch=build.branch,
            run_id=build.run_id,
        ))

        try:
            # Run the command
            process = await asyncio.create_subprocess_shell(
                build.command,
                cwd=str(self.repo_dir),
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )

            stdout, stderr = await process.communicate()

            build.output = stdout.decode() if stdout else ""
            build.error = stderr.decode() if stderr else ""
            build.return_code = process.returncode
            build.finished_at = datetime.now()

            if process.returncode == 0:
                build.status = BuildStatusType.SUCCESS
                await self.event_emitter.emit_build_success(
                    branch=build.branch,
                    run_id=build.run_id,
                )
            else:
                build.status = BuildStatusType.FAILURE
                await self.event_emitter.emit_build_failure(
                    branch=build.branch,
                    run_id=build.run_id,
                    error_message=build.error[:500],
                )

        except Exception as e:
            build.status = BuildStatusType.FAILURE
            build.error = str(e)
            build.finished_at = datetime.now()

            await self.event_emitter.emit_build_failure(
                branch=build.branch,
                run_id=build.run_id,
                error_message=str(e),
            )

        logger.info(f"Build {build.run_id} completed with status {build.status.value}")

    async def get_build_status(self, run_id: str) -> BuildStatus:
        """Get status of a local build."""
        build = self._builds.get(run_id)
        if not build:
            raise ValueError(f"Unknown build: {run_id}")

        return BuildStatus(
            run_id=build.run_id,
            status=build.status,
            branch=build.branch,
            started_at=build.started_at,
            finished_at=build.finished_at,
            error_message=build.error[:500] if build.error else None,
        )

    async def cancel_build(self, run_id: str) -> bool:
        """Cancel a local build (best effort)."""
        build = self._builds.get(run_id)
        if not build:
            return False

        if build.status == BuildStatusType.RUNNING:
            build.status = BuildStatusType.CANCELLED
            build.finished_at = datetime.now()
            logger.info(f"Cancelled build {run_id}")
            return True

        return False

    async def create_pr(
        self,
        source_branch: str,
        target_branch: str,
        title: str,
        body: str = "",
    ) -> PRInfo:
        """Create a simulated PR."""
        pr_number = self._next_pr_number
        self._next_pr_number += 1

        pr = LocalPR(
            pr_number=pr_number,
            source_branch=source_branch,
            target_branch=target_branch,
            title=title,
            body=body,
        )
        self._prs[pr_number] = pr

        # Emit PR created event
        await self.event_emitter.emit(CIEvent(
            event_id=str(uuid.uuid4()),
            event_type=CIEventType.PR_CREATED.value,
            branch=source_branch,
            pr_number=pr_number,
        ))

        logger.info(f"Created local PR #{pr_number}: {title}")

        # Auto-run checks
        asyncio.create_task(self._run_pr_checks(pr))

        return PRInfo(
            pr_number=pr_number,
            url=f"local://pr/{pr_number}",
            title=title,
            source_branch=source_branch,
            target_branch=target_branch,
            status=pr.status,
            created_at=pr.created_at,
        )

    async def _run_pr_checks(self, pr: LocalPR):
        """Run checks on a PR."""
        # Trigger a build for the PR
        run_id = await self.trigger_build(pr.source_branch)

        # Wait for completion
        while True:
            status = await self.get_build_status(run_id)
            if status.is_complete():
                pr.checks_passed = status.is_success()

                event_type = (
                    CIEventType.PR_CHECKS_PASSED
                    if pr.checks_passed
                    else CIEventType.PR_CHECKS_FAILED
                )
                await self.event_emitter.emit(CIEvent(
                    event_id=str(uuid.uuid4()),
                    event_type=event_type.value,
                    branch=pr.source_branch,
                    pr_number=pr.pr_number,
                ))
                break

            await asyncio.sleep(1)

    async def get_pr_status(self, pr_number: int) -> PRInfo:
        """Get status of a PR."""
        pr = self._prs.get(pr_number)
        if not pr:
            raise ValueError(f"Unknown PR: {pr_number}")

        return PRInfo(
            pr_number=pr.pr_number,
            url=f"local://pr/{pr_number}",
            title=pr.title,
            source_branch=pr.source_branch,
            target_branch=pr.target_branch,
            status=pr.status,
            created_at=pr.created_at,
            merged_at=pr.merged_at,
            checks_passed=pr.checks_passed,
        )

    async def merge_pr(self, pr_number: int) -> bool:
        """Merge a PR locally."""
        pr = self._prs.get(pr_number)
        if not pr:
            return False

        if pr.status != PRStatusType.OPEN:
            logger.warning(f"PR {pr_number} is not open")
            return False

        if not pr.checks_passed:
            logger.warning(f"PR {pr_number} checks have not passed")
            return False

        # Perform local git merge
        try:
            # Checkout target and merge source
            subprocess.run(
                ["git", "checkout", pr.target_branch],
                cwd=str(self.repo_dir),
                check=True,
                capture_output=True,
            )
            subprocess.run(
                ["git", "merge", pr.source_branch, "--no-ff", "-m",
                 f"Merge PR #{pr_number}: {pr.title}"],
                cwd=str(self.repo_dir),
                check=True,
                capture_output=True,
            )

            pr.status = PRStatusType.MERGED
            pr.merged_at = datetime.now()

            # Emit merged event
            await self.event_emitter.emit_pr_merged(
                branch=pr.source_branch,
                pr_number=pr_number,
            )

            logger.info(f"Merged PR #{pr_number}")
            return True

        except subprocess.CalledProcessError as e:
            logger.error(f"Failed to merge PR {pr_number}: {e.stderr}")
            # Abort merge
            subprocess.run(
                ["git", "merge", "--abort"],
                cwd=str(self.repo_dir),
                check=False,
            )
            return False

    async def subscribe(
        self,
        callback: Callable[[CIEvent], Awaitable[None]]
    ):
        """Subscribe to CI events."""
        self._subscribers.append(callback)
        await self.event_emitter.subscribe(callback)

    async def unsubscribe(
        self,
        callback: Callable[[CIEvent], Awaitable[None]]
    ):
        """Unsubscribe from CI events."""
        if callback in self._subscribers:
            self._subscribers.remove(callback)
        await self.event_emitter.unsubscribe(callback)

    # Local-specific methods

    def get_all_builds(self) -> List[LocalBuild]:
        """Get all builds (local-specific)."""
        return list(self._builds.values())

    def get_all_prs(self) -> List[LocalPR]:
        """Get all PRs (local-specific)."""
        return list(self._prs.values())

    def get_build_output(self, run_id: str) -> Optional[str]:
        """Get full build output (local-specific)."""
        build = self._builds.get(run_id)
        return build.output if build else None

    def get_build_error(self, run_id: str) -> Optional[str]:
        """Get full build error output (local-specific)."""
        build = self._builds.get(run_id)
        return build.error if build else None

File: src/ci/local.js
/**
 * @file Local CI provider implementation.
 * Provides git-based CI operations for local development.
 * @module ci/local
 */

import { spawn } from 'node:child_process';
import { writeFile, readFile, mkdir } from 'node:fs/promises';
import { join } from 'node:path';
import { CIProvider, BuildStatus, PRInfo } from './interface.js';
import { CIEventEmitter } from './events.js';
import { BuildStatusType, PRStatusType } from '../types/index.js';
import { CIError, TimeoutError } from '../orchestrator/errors.js';

/**
 * Utility to sleep for a given time.
 * @param {number} ms - Milliseconds to sleep
 * @returns {Promise<void>}
 */
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

/**
 * Local CI provider using git operations.
 * Suitable for local development and testing.
 * @extends CIProvider
 */
export class LocalCIProvider extends CIProvider {
  /**
   * Create a LocalCIProvider.
   * @param {Object} props - Properties
   * @param {string} props.repoDir - Repository directory
   * @param {string} [props.stateDir] - State directory for tracking builds/PRs
   * @param {string} [props.integrationBranch='integration'] - Integration branch name
   */
  constructor({ repoDir, stateDir, integrationBranch = 'integration' }) {
    super();
    /** @type {string} */
    this.repoDir = repoDir;
    /** @type {string} */
    this.stateDir = stateDir ?? join(repoDir, '.state', 'ci');
    /** @type {string} */
    this.integrationBranch = integrationBranch;
    /** @type {CIEventEmitter} */
    this.eventEmitter = new CIEventEmitter();
    /** @private @type {Map<string, BuildStatus>} */
    this._builds = new Map();
    /** @private @type {Map<number, PRInfo>} */
    this._prs = new Map();
    /** @private @type {number} */
    this._nextPRNumber = 1;
    /** @private @type {number} */
    this._nextRunId = 1;
  }

  /**
   * Run a git command.
   * @private
   * @param {string[]} args - Git command arguments
   * @returns {Promise<{stdout: string, stderr: string, code: number}>}
   */
  async _runGit(args) {
    return new Promise((resolve, reject) => {
      const process = spawn('git', args, { cwd: this.repoDir });

      let stdout = '';
      let stderr = '';

      process.stdout?.on('data', (data) => {
        stdout += data.toString();
      });

      process.stderr?.on('data', (data) => {
        stderr += data.toString();
      });

      process.on('error', (error) => {
        reject(new CIError(`Git command failed: ${error.message}`, {
          cause: error,
          provider: 'local',
          operation: args.join(' '),
        }));
      });

      process.on('close', (code) => {
        resolve({ stdout: stdout.trim(), stderr: stderr.trim(), code: code ?? 0 });
      });
    });
  }

  /**
   * Ensure state directory exists.
   * @private
   * @returns {Promise<void>}
   */
  async _ensureStateDir() {
    await mkdir(this.stateDir, { recursive: true });
  }

  /**
   * Trigger a build on a branch.
   * For local provider, this just validates the branch exists and marks build as running.
   *
   * @param {string} branch - Branch name
   * @returns {Promise<BuildStatus>}
   */
  async triggerBuild(branch) {
    await this._ensureStateDir();

    const runId = `local-${this._nextRunId++}`;

    // Verify branch exists
    const result = await this._runGit(['rev-parse', '--verify', branch]);
    if (result.code !== 0) {
      const status = new BuildStatus({
        runId,
        status: BuildStatusType.FAILURE,
        startedAt: new Date(),
        completedAt: new Date(),
        errorMessage: `Branch ${branch} not found`,
      });
      this._builds.set(runId, status);
      await this.eventEmitter.emitBuildFailure(runId, branch, status.errorMessage);
      return status;
    }

    const status = new BuildStatus({
      runId,
      status: BuildStatusType.RUNNING,
      startedAt: new Date(),
    });

    this._builds.set(runId, status);
    await this.eventEmitter.emitBuildStarted(runId, branch);

    // Simulate build completion after a short delay
    setTimeout(async () => {
      status.status = BuildStatusType.SUCCESS;
      status.completedAt = new Date();
      await this.eventEmitter.emitBuildSuccess(runId, branch);
    }, 100);

    return status;
  }

  /**
   * Get the status of a build.
   * @param {string} runId - Build run ID
   * @returns {Promise<BuildStatus>}
   */
  async getBuildStatus(runId) {
    const status = this._builds.get(runId);
    if (!status) {
      throw new CIError(`Build ${runId} not found`, {
        provider: 'local',
        operation: 'getBuildStatus',
      });
    }
    return status;
  }

  /**
   * Wait for a build to complete.
   * @param {string} runId - Build run ID
   * @param {number} [timeout=300000] - Timeout in milliseconds
   * @returns {Promise<BuildStatus>}
   */
  async waitForBuild(runId, timeout = 300000) {
    const startTime = Date.now();
    const pollInterval = 500;

    while (Date.now() - startTime < timeout) {
      const status = await this.getBuildStatus(runId);
      if (status.isComplete()) {
        return status;
      }
      await sleep(pollInterval);
    }

    throw new TimeoutError(`Build ${runId} timed out`, {
      operation: 'waitForBuild',
      timeoutMs: timeout,
    });
  }

  /**
   * Create a pull request.
   * For local provider, this creates a merge commit simulation.
   *
   * @param {Object} props - Properties
   * @param {string} props.title - PR title
   * @param {string} props.body - PR body
   * @param {string} props.sourceBranch - Source branch
   * @param {string} props.targetBranch - Target branch
   * @returns {Promise<PRInfo>}
   */
  async createPR({ title, body, sourceBranch, targetBranch }) {
    await this._ensureStateDir();

    const prNumber = this._nextPRNumber++;

    // Verify source branch exists
    const sourceResult = await this._runGit(['rev-parse', '--verify', sourceBranch]);
    if (sourceResult.code !== 0) {
      throw new CIError(`Source branch ${sourceBranch} not found`, {
        provider: 'local',
        operation: 'createPR',
      });
    }

    const prInfo = new PRInfo({
      number: prNumber,
      title,
      status: PRStatusType.OPEN,
      url: `local://pr/${prNumber}`,
      sourceBranch,
      targetBranch,
    });

    this._prs.set(prNumber, prInfo);

    // Save PR info to state file
    const prPath = join(this.stateDir, `pr-${prNumber}.json`);
    await writeFile(prPath, JSON.stringify({
      ...prInfo.toDict(),
      body,
    }, null, 2));

    await this.eventEmitter.emitPROpened(prNumber, sourceBranch);
    console.log(`[LocalCI] Created PR #${prNumber}: ${title}`);

    return prInfo;
  }

  /**
   * Get the status of a pull request.
   * @param {number} prNumber - PR number
   * @returns {Promise<PRInfo>}
   */
  async getPRStatus(prNumber) {
    let prInfo = this._prs.get(prNumber);

    if (!prInfo) {
      // Try to load from state file
      const prPath = join(this.stateDir, `pr-${prNumber}.json`);
      try {
        const content = await readFile(prPath, 'utf-8');
        const data = JSON.parse(content);
        prInfo = PRInfo.fromDict(data);
        this._prs.set(prNumber, prInfo);
      } catch {
        throw new CIError(`PR #${prNumber} not found`, {
          provider: 'local',
          operation: 'getPRStatus',
        });
      }
    }

    return prInfo;
  }

  /**
   * Merge a pull request.
   * For local provider, this performs an actual git merge.
   *
   * @param {number} prNumber - PR number
   * @returns {Promise<PRInfo>}
   */
  async mergePR(prNumber) {
    const prInfo = await this.getPRStatus(prNumber);

    if (prInfo.isMerged()) {
      return prInfo;
    }

    if (!prInfo.isOpen()) {
      throw new CIError(`PR #${prNumber} is not open`, {
        provider: 'local',
        operation: 'mergePR',
      });
    }

    // Checkout target branch
    await this._runGit(['checkout', prInfo.targetBranch]);

    // Merge source branch
    const mergeResult = await this._runGit([
      'merge',
      prInfo.sourceBranch,
      '--no-edit',
      '-m',
      `Merge PR #${prNumber}: ${prInfo.title}`,
    ]);

    if (mergeResult.code !== 0) {
      throw new CIError(`Failed to merge PR #${prNumber}: ${mergeResult.stderr}`, {
        provider: 'local',
        operation: 'mergePR',
      });
    }

    // Update PR status
    prInfo.status = PRStatusType.MERGED;
    prInfo.mergedAt = new Date();

    // Save updated PR info
    const prPath = join(this.stateDir, `pr-${prNumber}.json`);
    await writeFile(prPath, JSON.stringify(prInfo.toDict(), null, 2));

    await this.eventEmitter.emitPRMerged(prNumber, prInfo.sourceBranch);
    console.log(`[LocalCI] Merged PR #${prNumber}`);

    return prInfo;
  }

  /**
   * Wait for a PR to be merged.
   * @param {number} prNumber - PR number
   * @param {number} [timeout=600000] - Timeout in milliseconds
   * @returns {Promise<PRInfo>}
   */
  async waitForPRMerge(prNumber, timeout = 600000) {
    const startTime = Date.now();
    const pollInterval = 1000;

    while (Date.now() - startTime < timeout) {
      const prInfo = await this.getPRStatus(prNumber);
      if (prInfo.isMerged()) {
        return prInfo;
      }
      if (prInfo.isClosed()) {
        throw new CIError(`PR #${prNumber} was closed without merge`, {
          provider: 'local',
          operation: 'waitForPRMerge',
        });
      }
      await sleep(pollInterval);
    }

    throw new TimeoutError(`PR #${prNumber} merge timed out`, {
      operation: 'waitForPRMerge',
      timeoutMs: timeout,
    });
  }

  /**
   * Subscribe to CI events.
   * @param {import('./interface.js').CIEventHandler} handler - Event handler
   * @returns {Promise<void>}
   */
  async subscribe(handler) {
    this.eventEmitter.subscribe(handler);
  }

  /**
   * Unsubscribe from CI events.
   * @param {import('./interface.js').CIEventHandler} handler - Event handler
   * @returns {Promise<void>}
   */
  async unsubscribe(handler) {
    this.eventEmitter.unsubscribe(handler);
  }

  /**
   * Get all tracked PRs.
   * @returns {PRInfo[]}
   */
  getAllPRs() {
    return Array.from(this._prs.values());
  }

  /**
   * Get all tracked builds.
   * @returns {BuildStatus[]}
   */
  getAllBuilds() {
    return Array.from(this._builds.values());
  }
}
